build_modules/lib/src/module_library.dart

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/analyzer.dart';
+import 'package:build/build.dart';
+import 'package:meta/meta.dart';
+
+class ModuleLibrary {
+  /// The AssetId of the original Dart source file.
+  final AssetId id;
+
+  /// Whether this library can be imported.
+  ///
+  /// This will be false if the source file is a "part of", or imports code that
+  /// can't be used outside the SDK.
+  final bool isImportable;
+
+  /// Whether this library is an entrypoint.
+  ///
+  /// True if the library is in `lib/` but not `lib/src`, or if it is outside of
+  /// `lib/` and contains a `main` method. Always false if this is not an
+  /// importable library.
+  final bool isEntryPoint;
+
+  /// The IDs of libraries that are imported or exported by this library.
+  ///
+  /// Null if this is not an importable library.
+  final Set<AssetId> deps;
+
+  /// The "part" files for this library.
+  ///
+  /// Null if this is not an importable library.
+  final Set<AssetId> parts;
+
+  ModuleLibrary._(this.id,
+      {@required this.isEntryPoint, @required this.deps, @required this.parts})
+      : isImportable = true;
+
+  ModuleLibrary._nonImportable(this.id)
+      : isImportable = false,
+        isEntryPoint = false,
+        deps = null,
+        parts = null;
+
+  factory ModuleLibrary._fromCompilationUnit(
+      AssetId id, bool isEntryPoint, CompilationUnit parsed) {
+    var deps = new Set<AssetId>();
+    var parts = new Set<AssetId>();
+    for (var directive in parsed.directives) {
+      if (directive is! UriBasedDirective) continue;
+      var path = (directive as UriBasedDirective).uri.stringValue;
+      if (Uri.parse(path).scheme == 'dart') continue;
+      var linkedId = new AssetId.resolve(path, from: id);
+      if (linkedId == null) continue;
+      if (directive is PartDirective) {
+        parts.add(linkedId);
+        continue;
+      }
+
+      List<Configuration> conditionalDirectiveConfigurations;
+
+      if (directive is ImportDirective && directive.configurations.isNotEmpty) {
+        conditionalDirectiveConfigurations = directive.configurations;
+      } else if (directive is ExportDirective &&
+          directive.configurations.isNotEmpty) {
```

Comment:
When can directive.configurations be non empty if it isn't an import or export directive? If it can't, then can we get rid of the check?

Reply:
This is https://github.com/dart-lang/build/pull/1649/files#diff-e6beb981c2249722d51e60e34b104f88L97

It happens with config specific imports. When we add support for different modules based on platform this line will need to change - we'll need to consider only the the configuration that matches the platform.

Reply:
I don't follow. Maybe I didn't phrase my question clearly. Do we even need to check if `directive is ImportDirective` or `directive is ExportDirective`? My assumption is that `directive.configurations.isNotEmpty` can only be true when the directive is one of those two types. Thus the logic can be simplified to:

```
if (directive.configurations.isNotEmpty) {
   conditionalDirectiveConfigurations = directive.configurations;
}
```

Reply:
For some context on the code: https://github.com/dart-lang/build/pull/1524

The `.configurations` field only exists on [NamespaceDirective](https://pub.dartlang.org/documentation/analyzer/latest/analyzer.dart.ast.ast/NamespaceDirective-class.html), not on `UriBasedDirective` so we need check at least that. We could probably check only that and skip the separate branches for `ImportDirective` and `ExportDirective` though, so I'll take a look at that in a followup.